Scripts/Runtime/Collections/DeferredNativeArray.cs

```diff
@@ -0,0 +1,333 @@
+using System;
+using System.Diagnostics;
+using System.Runtime.InteropServices;
+using Unity.Burst;
+using Unity.Collections;
+using Unity.Collections.LowLevel.Unsafe;
+using Unity.Jobs;
+
+namespace Anvil.Unity.DOTS.Collections
+{
+    [BurstCompatible]
+    [StructLayout(LayoutKind.Sequential)]
+    internal unsafe struct DeferredNativeArrayBufferInfo
+    {
+        public static readonly int SIZE = UnsafeUtility.SizeOf<DeferredNativeArrayBufferInfo>();
+        public static readonly int ALIGNMENT = UnsafeUtility.AlignOf<DeferredNativeArrayBufferInfo>();
+
+        [NativeDisableUnsafePtrRestriction] public void* Buffer;
+        public int Length;
+        public int MaxIndex;
+        public DeferredNativeArrayState State;
+    }
+
+    internal enum DeferredNativeArrayState
+    {
+        Placeholder,
```

Comment:
`IsCreated` refers to whether the collection has had any allocations because if it has (even the initial temp one) we need to dispose it to clear the memory. So we already have that which checks if the `m_BufferInfo` and `Buffer` pointers exist.

Because it's a struct you could have an array of `DeferredNativeArray`s but no memory will be allocated because you can't have a default constructor (yet) in C#. 

All these native collections have the same issue because of Burst/Jobs so Unity's solution is to use a struct that wraps a pointer but then must have the `IsCreated` to see if it was actually created so we know to clear the memory. 

Anyway, all that to say, I realized that I can just remove the enum all together because I can use the `Length` instead. 

At the beginning the `Length` is 0 and after you deferred create it will be greater than 0.

Reply:
Good point. Fixed.